File: kickstart/2022/coding_practice_with_kick_start_session_3/sample_problem/problem_statement/samples/sample.ts
declare var require: any;
declare var process: any;

const readline = require('readline');
let rl = readline.createInterface(process.stdin, process.stdout);

// Define an interface for the data needed for each test case.
interface TestData {
  testNumber: number;
  numKids: number;
  candyBags: number[];
}

function parseInput(input: string[]) {
  let line: number = 0;
  // Read the number of test cases from the first line.
  const numCases = Number(input[line++]);

  const testData: TestData[] = [];
  for (let testNumber = 1; testNumber <= numCases; ++testNumber) {
    // The first line of each test case contains the number of candyBags
    // and the number of kids. The first number isn't needed.
    const numKids = Number(input[line++].split(' ')[1]);

    // The second line of each test contains the number of candies in each
    // bag.
    const candyBags = input[line++].split(' ').map((str) => Number(str));

    testData.push({testNumber, numKids, candyBags});
  }
  return testData;
}

function runTestCase(data: TestData) {
  // Compute the total number of candies.
  const sum = data.candyBags.reduce((sum, candies) => sum + candies, 0);

  // Compute the remainder after distributing the candies equally.
  const remainder = sum % data.numKids;

  // Print the result
  console.log(`Case #${data.testNumber}: ${remainder}`);
}

function runAllTests(input: string[]) {
  // Parse the input into TestData objects.
  const testCases = parseInput(input);

  // Run each test case.
  testCases.forEach(runTestCase);
}

// Read the input data and run all test cases.
const lines: string[] = [];
rl.on('line', line => lines.push(line)).on('close', () => runAllTests(lines));

File: kickstart/2022/coding_practice_with_kick_start_session_3/sample_problem/problem_statement/samples/sample.ml
let solve n m c =
  (* Calculate the total number of candies *)
  let total = Array.fold_left (+) 0 c in
  (* Calculate and return the division remainder *)
  total mod m

let main =
  (* Read the number of tests *)
  let t = Scanf.scanf "%d" (fun t -> t) in
  (* For each test *)
  for test = 1 to t do
    (* Read the number of bags and kids *)
    let n, m = Scanf.scanf " %d %d" (fun n m -> n, m) in
    (* Read the number of candies per bag *)
    let c = Array.init n (fun _ -> Scanf.scanf " %d" (fun h -> h)) in
    (* Solve the task *)
    let result = solve n m c in
    (* Print the result *)
    Printf.printf "Case #%d: %d\n" test result
  done

